src/datepicker/types.js

```diff
@@ -223,101 +233,78 @@ export type SharedStylePropsT = {
 
 export type StateChangeTypeT = ?$Values<typeof STATE_CHANGE_TYPE>;
 
-export type ContainerStateT = {
+export type ContainerStateT<T = Date> = {
   /** Selected `Date`. If `range` is set, `value` is an array of 2 values. */
-  value?: ?Date | Array<Date>,
+  value?: ?T | Array<T>,
 };
 
-export type NavigationContainerStateT = {
+export type NavigationContainerStateT<T = Date> = {
   // indicates a highlighted date on hover and keyboard navigation
-  highlightedDate?: ?Date,
+  highlightedDate?: ?T,
   // used to disable keyboard navigation when a month or year select
   // dropdown is opened
   isActive?: boolean,
   // last remembered highlighted date to restore
   // when keyboard navigating after a mouse moved off the cal and reset
   // highlightedDate value
-  lastHighlightedDate?: Date,
+  lastHighlightedDate?: T,
 };
 
-export type StateReducerT = (
+export type StateReducerT<T = Date> = (
   stateType: StateChangeTypeT,
-  nextState: ContainerStateT,
-  currentState: ContainerStateT,
-) => ContainerStateT;
+  nextState: ContainerStateT<T>,
+  currentState: ContainerStateT<T>,
+) => ContainerStateT<T>;
 
-export type NavigationContainerStateReducerT = (
+export type NavigationContainerStateReducerT<T = Date> = (
   stateType: StateChangeTypeT,
-  nextState: NavigationContainerStateT,
-  currentState: NavigationContainerStateT,
-) => NavigationContainerStateT;
+  nextState: NavigationContainerStateT<T>,
+  currentState: NavigationContainerStateT<T>,
+) => NavigationContainerStateT<T>;
 
-export type StatefulContainerPropsT<T> = {
-  children: T => React.Node,
+export type StatefulContainerPropsT<PropsT, T = Date> = {
```

Comment:
I had to rename the generic here so it didn't conflict with the new generic date type